docs/rfcs/2025-04-14-storage-keys.md

```diff
@@ -0,0 +1,244 @@
+
+Neon will interoperate with different KMS APIs on different platforms.  We will implement a generic interface,
+similar to how `remote_storage` wraps different object storage APIs:
+- `generate(accountId, keyType, alias) -> `
+- `unwrap(accountId, ciphertext key) -> plaintext key`
+
+Hereafter, when we talk about generating or unwrapping a key, this means a call into the KMS API.
+
+The KMS deals with abstract "account IDs", which are not equal to tenant IDs and may not be
+1:1 with tenants.  The account ID will be provided as part of tenant configuration, along
+with a field to identify an encryption mode.
+
+### Pageserver key storage
+
+The wrapped pageserver timeline key will be stored in the timeline index object.  Because of
+key rotation, multiple keys will be stored in an array, with each key having a counter version. 
+
+```
+"keys": [
+    {
+        # The key version: a new key with the next version is generated when rekeying
+        "version": 1,
+        # The wrapped key: this is unwrapped by a KMS API call when the key is to be used
+        "wrapped": "<base64 string>",
+        # The time the key was generated: this may be used to implement rekeying/key rotation
+        # policies.
+        "ctime": "<ISO 8601 timestamp>",
+    },
+    ...
+]
+```
+
+Wrapped pageserver layer file keys will be stored in the `index_part` file, as part
+of the layer metadata.
+
+```
+# LayerFileMetadata
+{
+    "key": {
+        "version":
+    }
+
+}
+```
+
+To enable re-key procedure to drop deleted versions with old keys, and to avoid mistakes in index_part leading to irretreivable data loss, wrapped keys & version will also be stored
+in the object store metadata of uploaded objects.
+
+### Safekeeper key storage
+
+All safekeeper storage is per-timeline.  The only concept of a tenant in the safekeeper
+is as a namespace for timelines.
+
+As the safekeeper doesn't currently have a flexible metadata object in remote storage,
+we will add one.  This will initially contain:
+- A configuration object that contains the accountId
+- An array of keys idential to those used in the pageserver's index.
+
+Because multiple safekeeper processes share the same remote storage path, we must be
+sure to handle write races safely.  To avoid giving safekeepers a pageserver-like generation
+concept (not to be confused with safekeeper's configuration generation), we may use
+the conditional write primitive that is available on S3 and ABS, to implement a safe
+read-then-write for operations such as key rotation, such that a given key version is
+only ever implemented once.
+
+### Key rotation
+
+The process of key rotation is:
+1. Load the version of the existing key
+2. Generate a new key
+3. Store the new key with the previous version incremented by 1
+4. **Only once durably stored** use the new key for subsequent generation of object keys
+
+This is the same for safekeepers and pageservers.
+
+A storage controller API will be exposed for re-keying.
```

Comment:
the section right after makes a distinction between key rotation and re-keying but this section seems to use them synonymously.

Reply:
oops, typo